_posts/2019-07-01-test-design.md

```diff
@@ -0,0 +1,294 @@
+
+This is very convenient for most tests.
+However, in some tests we want to influence how the Order is constructed, so we can test some custom cases other than a default scenario, specific for certain tests.
+We implemented this by adding some methods to our scenario class in that allow the scenario to be modified to the test's needs.
+
+```java
+public class CreateOrderScenario {
+
+    public static final UUID customerId = UUID.randomUUID();
+
+    private CreateOrderData createOrderData;
+    private CustomerRepository customerRepository;
+
+    public static CreateOrderScenario defaultScenario() {
+        final CreateOrderScenario scenario = new CreateOrderScenario();
+
+        final Customer customer = mock(Customer.class);
+        when(customer.isActive()).thenReturn(true);
+
+        scenario.customerRepository = mock(CustomerRepository.class);
+        when(scenario.customerRepository.findById(customerId)).thenReturn(customer);
+
+        scenario.createOrderData = new CreateOrderData(customerId);
+
+        return scenario;
+    }
+
+    public CreateOrderScenario modifyCreateOrderData(Consumer<CreateOrderData> modifier) {
+        modifier.accept(createOrderData);
+        return this;
+    }
+
+    public CreateOrderScenario overrideCustomerRepository(Consumer<OrderRepository> modifier) {
+        customerRepository = mock(CustomerRepository.class);
+        modifier.accept(customerRepository);
+        return this;
+    }
+
+    public Order execute() {
+        final OrderValidator orderValidator = new OrderValidator(customerRepository);
+        final OrderFactory orderFactory = new OrderFactory(orderValidator);
+
+        return orderFactory.createOrder(createOrderData);
+    }
+}
+```
+
+In the following example we created two tests that verify that the construction of an order works correctly.
+In the first test, we use the default scenario without modifying anything, meaning, we test the happy path and verify that all data in the created order is correct.
+
+In the second test we verify that if we try to create an order for a customer that doesn't exist, we get a validation exception.
+We do this by creating a default scenario, then modifying the input data to use a customerId defined in the test, and then overriding the behaviour of the CustomerRepository mock.
+
+```java
+class CreateOrderTest {
+    
+    @Test
+    void createOrder_happyPath() {
+        final Order order = CreateOrderScenario
+            .defaultScenario()
+            .execute();
+        
+        assertThat(order.getCustomerId()).isEqualTo(CreateOrderScenario.customerId);
+    }
+    
+    @Test
+    void createOrder_customerInactive() {
+        final UUID customerId = UUID.randomUUID();
+        final CreateOrderScenario scenario = CreateOrderScenario
+            .defaultScenario()
+            .modifyCreateOrderData(orderData -> orderData.setCustomerId(customerId))
+            .overrideCustomerRepository(repository -> 
```

Comment:
In this case overriding the mocking shouldn't be necessary as the default scenario specifically mocks the `findById` call with a specific `customerId`. So all calls with a different `customerId` will result in a `null` by default.
But I understand the example you want to make. :)

Reply:
fixed by letting the repository return an optional ;)